libraries/base/GHC/IO/Handle.hs

```diff
@@ -639,7 +639,11 @@ dupHandle_ :: (IODevice dev, BufferedIO dev, Typeable dev) => dev
 dupHandle_ new_dev filepath other_side h_@Handle__{..} mb_finalizer = do
    -- XXX wrong!
   mb_codec <- if isJust haEncoder then fmap Just getLocaleEncoding else return Nothing
-  mkHandle new_dev filepath haType True{-buffered-} mb_codec
+  is_tty <- IODevice.isTerminal new_dev
+  let buffer_mode
+         | is_tty = LineBuffering
+         | otherwise = BlockBuffering Nothing
+  mkHandle new_dev filepath haType buffer_mode mb_codec
```

Comment:
Another alternative to implement this is to make `Maybe BufferMode` as one of the argument to `mkHandle` and a value of `Nothing` should do the `isTerminal` check. I think doing it this way would make the code a lot more concise and avoid the repetitive `isTerminal` checks.